performance: prefetch prod_type + open finding list, prod list, prod engagements

valentijnscholten opened the pull request:
We currently have about 120 products, 400 engagements, 1200 findings. In this scenario Defect Dojo becomes quite slow, mainly because it suffers from the [1+N queries problem](https://scoutapm.com/blog/finding-and-fixing-n-1s-in-django-apps).

The best way to fix this is to make queries more eager loading by using `select_related`, `prefetch_related` and `Prefetch()`. Also moving some of the counting to the database will help.

This PR is a first step to implement these steps for:
- list of prod_types
- list of products
- list of engagements on the product details page
- list of open findings (product/engagement/global)

It took a bit of fiddling to get it all working, especially since we're also using paging.

I think there are lots more places where the above could help. This PR is a first step to get some feedback and also make sure the PR does not become 'too big to review'.

Please let me know your thoughts. 

I have been running most of this for a couple of weeks now and it really helps performance.
Some pages required 350 queries before and are now reduced to 150. These 150 are mainly around system settings, which will get an improvement in #1868.

In the end Defect Dojo could also benefit from an (ORM) caching solution. But even with a caching solution, the N+1 queries problem needs to be gone to have good performance.

madchap commented:
fyi @jvz

valentijnscholten commented:
Anyone know why travis keeps failing here?

valentijnscholten commented:
It looks like they are all failing. Or are they all failing due to the integration test failing? Let me try and run the integration tests locally then.

Maffooch commented:
There’s an issue with migrations. It’s trying to run an auto generated migration. Check the logs of the Travis test for more details.

valentijnscholten commented:
> Looks like an amazing improvement! Nice use of ORM features. :)
Yeah, was quite a nice learning journey for someone who didn't knew python, django, django-orm, django-filter, ...

> There’s an issue with migrations. It’s trying to run an auto generated migration. Check the logs of the Travis test for more details.
Ah, of course. At first I thought it was just travis having glitches again without realizing the raw log would be helpful. Looks like django thinks it needs to apply migrations because I added 'related_name' in two places. I think I can tweak the queries a bit to get rid of them so no migrations are needed. (They shouldn't be needed really because related name doesn't affect the database model).

valentijnscholten commented:
Solved, but it keeps failing with
```
selenium.common.exceptions.NoSuchElementException: Message: Could not locate element with visible text: Research and Development
```
Normally it's just a matter of restarting the job a couple of times to get it in a passing state, but now  it seems to keep failing. Anyone knows why? I can't get the unit tests running locally let alone travis.

ptrovatelli commented:
> Solved, but it keeps failing with
> 
> ```
> selenium.common.exceptions.NoSuchElementException: Message: Could not locate element with visible text: Research and Development
> ```
> 
> Normally it's just a matter of restarting the job a couple of times to get it in a passing state, but now it seems to keep failing. Anyone knows why? I can't get the unit tests running locally let alone travis.

the docker "unit_tests" env will not help you with the selenium test. it's not running them. only way I've found to run the selenium test is a real ubuntu env. 

i'd started working on a dockerised env for this but it wasn't as practical as a real ubuntu so i've let it down.

ptrovatelli commented:
other flavors of linux should work too. no need to install a complete travis locally; tried this, was a real pain. just a linux with selenium and a couple of other stuff that travis installs before running the selenium test (chrome driver and stuff).

valentijnscholten commented:
> the docker "unit_tests" env will not help you with the selenium test. it's not running them. only way I've found to run the selenium test is a real ubuntu env.
I know, but at least unit_tests are something that (in theory) I should be able to run easily locally. So I thought let's make sure those are running fine first.

valentijnscholten commented:
ok, think I found it. The `test_delete_product_type` deletes the wrong product type. It deletes the first one in the list, which usually is the one it just had created itself. however due to the prefetching it seems the ordering has changed and now it deletes the `Research and Development`  product_type.

valentijnscholten commented:
Solved everything, build is now passing also in Travis. Just squashed everything into 1 commit to get rid of all the intermediate debugging steps. I added default sorting for the ProductType model, which means there is a new migration in the PR now.

ptrovatelli commented on dojo/finding/views.py:
```diff
@@ -513,7 +516,7 @@ def open_findings(request, pid=None, eid=None, view=None):
     found_by = None
     try:
-        found_by = findings.found_by.all().distinct()
+        found_by = findings_filter.found_by.all().distinct()
```
why do we work on the findings_filter here instead of findings?

valentijnscholten replied:
because that is what the old code was also doing. findings doesn't have an attribute/property called found_by.

valentijnscholten commented on dojo/finding/views.py:
```diff
@@ -478,22 +479,24 @@ def open_findings(request, pid=None, eid=None, view=None):
             findings = Finding.objects.filter(active=True, duplicate=False).order_by('numerical_severity')
 
-    if request.user.is_staff:
-        findings = OpenFingingSuperFilter(
```
I don't understand, findings and findings_filter are available in both cases? For non staff there's just an extra `.filter`

ptrovatelli replied:
sorry you're right it's fine

valentijnscholten commented on dojo/product/views.py:
```diff
@@ -28,45 +28,67 @@
+    # otherwise the paginator will perform all the annotations/prefetching already only to count the total number of records
+    # see https://code.djangoproject.com/ticket/23771 and https://code.djangoproject.com/ticket/25375
+    name_words = [product.name for product in prods
+                    for word in product.name.split() if len(word) > 2]
```
No violations in my flake nor in travis. What are the violations?

valentijnscholten commented on dojo/product/views.py:
```diff
@@ -28,45 +28,67 @@
 
 
+def prefetch_for_product(prods):
+    prefetched_prods = prods.select_related('technical_contact').select_related('product_manager').select_related('prod_type').select_related('team_manager')
```
travis has some exceptions on which rules it skips, but my local flake8 also doesn't complain.

valentijnscholten commented:
> couldn't test yet; i'm having the "Too many levels of symbolic links" error.
> a couple of comments already though

yeah looks like something changed, maybe we have a dependency that is not frozen to a certain version and now a newer version is being used and causing issues on all our PRs.

valentijnscholten commented:
build errors should be fixed by #1937

ptrovatelli commented:
OK for me if travis passes; it seems hung?
